src/Voron/Impl/EncryptionBuffersPool.cs

```diff
@@ -1,125 +1,290 @@
+using Sparrow.Json.Parsing;
 using Sparrow.LowMemory;
 using Sparrow.Server;
 using Sparrow.Server.Platform;
 using Sparrow.Threading;
 using Sparrow.Utils;
-using Constants = Voron.Global.Constants;
 
 namespace Voron.Impl
 {
-    public unsafe class EncryptionBuffersPool : IDisposable, ILowMemoryHandler
+    public unsafe class EncryptionBuffersPool : ILowMemoryHandler
     {
-        private class NativeAllocation
-        {
-            public IntPtr Ptr;
-            public long Size;
-            public NativeMemory.ThreadStats AllocatingThread;
-        }
+        public static EncryptionBuffersPool Instance = new EncryptionBuffersPool();
 
+        private readonly long _maxBufferSizeToKeepInBytes = new Size(8, SizeUnit.Megabytes).GetValue(SizeUnit.Bytes);
+        private readonly MultipleUseFlag _lowMemoryFlag = new MultipleUseFlag();
         private readonly ConcurrentStack<NativeAllocation>[] _items;
-        private readonly DisposeOnce<SingleAttempt> _disposeOnceRunner;
+        private readonly Timer _cleanupTimer;
+        private DateTime _lastAllocationsRebuild = DateTime.UtcNow;
+        private long _generation;
+        private bool _hasDisposedAllocations;
+
+        public long Generation => _generation;
 
         public EncryptionBuffersPool()
         {
-            _items = new ConcurrentStack<NativeAllocation>[32];
+            var numberOfSlots = Bits.MostSignificantBit(_maxBufferSizeToKeepInBytes) + 1;
+            _items = new ConcurrentStack<NativeAllocation>[numberOfSlots];
+
             for (int i = 0; i < _items.Length; i++)
             {
                 _items[i] = new ConcurrentStack<NativeAllocation>();
             }
 
             LowMemoryNotification.Instance.RegisterLowMemoryHandler(this);
 
-            _disposeOnceRunner = new DisposeOnce<SingleAttempt>(() =>
-            {
-                ReleaseUnmanagedResources();
-                GC.SuppressFinalize(this);
-            });
+            _cleanupTimer = new Timer(CleanupTimer, null, TimeSpan.FromMinutes(1), TimeSpan.FromMinutes(1));
         }
 
         public byte* Get(int size, out NativeMemory.ThreadStats thread)
         {
             size = Bits.PowerOf2(size);
 
-            if (size > Constants.Size.Megabyte * 16)
+            if (size > _maxBufferSizeToKeepInBytes)
             {
                 // We don't want to pool large buffers
                 return PlatformSpecific.NativeMemory.Allocate4KbAlignedMemory(size, out thread);
             }
 
             var index = Bits.MostSignificantBit(size);
-
-            if (_items[index].TryPop(out var allocation))
+            while (_items[index].TryPop(out var allocation))
             {
-                thread = allocation.AllocatingThread;
-                return (byte*)allocation.Ptr;
+                if (allocation.InUse.Raise() == false)
+                    continue;
+
+                thread = NativeMemory.ThreadAllocations.Value;
+                thread.Allocations += size;
```

Comment:
This is wrong, I think.
You have _already_ allocated this, so you are double counting it, no?

Reply:
I update the allocations when the buffer is released to the pool.